pandar_pointcloud/src/lib/decoder/pandar40_decoder.cpp

```diff
@@ -86,81 +94,100 @@ void Pandar40Decoder::unpack(const pandar_msgs::PandarPacket& raw_packet)
   PointcloudXYZIRADT block_pc(new pcl::PointCloud<PointXYZIRADT>);
-  double unix_second = static_cast<double>(timegm(&packet_.t));
 
-  auto head = block_id + ((return_mode_ == ReturnMode::STRONGEST) ? 1 : 0);
-  auto tail = block_id + ((return_mode_ == ReturnMode::LAST) ? 1 : 2);
+  int even_block_id = block_id;
+  int odd_block_id = block_id + 1;
+  const auto& even_block = packet_.blocks[even_block_id];
+  const auto& odd_block = packet_.blocks[odd_block_id];
 
   for (auto unit_id : firing_order_) {
-    for (int i = head; i < tail; ++i) {
-      PointXYZIRADT point;
-      const auto& block = packet_.blocks[i];
-      const auto& unit = block.units[unit_id];
-      // skip invalid points
-      if (unit.distance <= 0.1 || unit.distance > 200.0) {
-        continue;
-      }
-      double xyDistance = unit.distance * cosf(deg2rad(elev_angle_[unit_id]));
-
-      point.x = static_cast<float>(
-          xyDistance * sinf(deg2rad(azimuth_offset_[unit_id] + (static_cast<double>(block.azimuth)) / 100.0)));
-      point.y = static_cast<float>(
-          xyDistance * cosf(deg2rad(azimuth_offset_[unit_id] + (static_cast<double>(block.azimuth)) / 100.0)));
-      point.z = static_cast<float>(unit.distance * sinf(deg2rad(elev_angle_[unit_id])));
-
-      point.intensity = unit.intensity;
-      point.distance = unit.distance;
 
-      point.ring = unit_id;
-      point.azimuth = block.azimuth;
+    const auto& even_unit = even_block.units[unit_id];
+    const auto& odd_unit = odd_block.units[unit_id];
 
-      point.time_stamp = unix_second + (static_cast<double>(packet_.usec)) / 1000000.0;
-      point.time_stamp -= (static_cast<double>(block_offset_dual_[i] + firing_offset_[unit_id]) / 1000000.0f);
+    bool even_usable = (even_unit.distance <= 0.1 || even_unit.distance > 200.0) ? 0 : 1;
+    bool odd_usable = (odd_unit.distance <= 0.1 || odd_unit.distance > 200.0) ? 0 : 1;  
 
-      block_pc->push_back(point);
+    if (return_mode_ == ReturnMode::STRONGEST) {
+      // Strongest return is in even block when both returns coincide
+      if (even_unit.intensity >= odd_unit.intensity && even_usable) {
+        block_pc->push_back(build_point(even_block_id, unit_id, ReturnType::SINGLE_STRONGEST));        
+      }
+      else if (even_unit.intensity < odd_unit.intensity && odd_usable) {
+        block_pc->push_back(build_point(odd_block_id, unit_id, ReturnType::SINGLE_STRONGEST)); 
+      }      
+    }
+    else if (return_mode_ == ReturnMode::LAST && even_usable) {
+      // Last return is always in even block
+      block_pc->push_back(build_point(even_block_id, unit_id, ReturnType::SINGLE_LAST)); 
+    }
+    else if (return_mode_ == ReturnMode::DUAL) {
+      // Only one return when returns are equal
+      if (even_unit.distance == odd_unit.distance && even_usable) {
+        block_pc->push_back(build_point(even_block_id, unit_id, ReturnType::DUAL_ONLY));
+      }
```

Comment:
`distance` is a floating point type, is there any problem to compare it without using epsilon?

Reply:
There is no problem without using epsilon as the float is generated from identical values.
But thinking about it, we probably want to discard echoes that are too close since upon investigation, close returns are not filtered (unlike Velodyne). There is no information about this in the datasheet, but sometimes echoes less than 10cm apart do appear (and are not particularly meaningful). I propose setting the threshold the same as Velodyne at 1m.

Reply:
I have added a parameter so we can adjust the threshold before dual returns at close distances will be considered a single return. The default is set to 0.1m, which is much smaller than Velodyne's 1m.